builder/job.go

```diff
@@ -121,24 +124,52 @@ func Build(d *daemon.Daemon, buildConfig *Config) error {
 	} else if urlutil.IsURL(buildConfig.RemoteURL) {
 		f, err := httputils.Download(buildConfig.RemoteURL)
 		if err != nil {
-			return err
+			return fmt.Errorf("Error downloading remote context %s: %s\n", buildConfig.RemoteURL, err)
 		}
 		defer f.Body.Close()
-		dockerFile, err := ioutil.ReadAll(f.Body)
-		if err != nil {
-			return err
+
+		plen := int(f.ContentLength)
+		if plen > maxPreambleLength {
```

Comment:
I now see that later on you'll throw an error if plen is -1 in DetectContentType(), I don't think that's correct since -1 is valid.